from pydantic import BaseModel


class VEP106BasicModel(BaseModel):

    Uploaded_variation: str
    Location: str
    Allele: str
    Gene: str
    Feature: str
    Feature_type: str
    Consequence: str
    cDNA_position: str
    CDS_position: str
    Protein_position: str
    Amino_acids: str
    Codons: str
    Existing_variation: str
    IMPACT: str
    DISTANCE: str
    STRAND: str
    FLAGS: str
    REFSEQ_MATCH: str
    SOURCE: str
    REFSEQ_OFFSET: str
    GIVEN_REF: str
    USED_REF: str
    BAM_EDIT: str


class VEP106FullModel(BaseModel):

    Uploaded_variation: str
    Location: str
    Allele: str
    Gene: str
    Feature: str
    Feature_type: str
    Consequence: str
    cDNA_position: str
    CDS_position: str
    Protein_position: str
    Amino_acids: str
    Codons: str
    Existing_variation: str
    IMPACT: str
    DISTANCE: str
    STRAND: str
    FLAGS: str
    VARIANT_CLASS: str
    SYMBOL: str
    SYMBOL_SOURCE: str
    HGNC_ID: str
    BIOTYPE: str
    CANONICAL: str
    MANE_SELECT: str
    MANE_PLUS_CLINICAL: str
    TSL: str
    APPRIS: str
    CCDS: str
    ENSP: str
    SWISSPROT: str
    TREMBL: str
    UNIPARC: str
    UNIPROT_ISOFORM: str
    REFSEQ_MATCH: str
    SOURCE: str
    REFSEQ_OFFSET: str
    GIVEN_REF: str
    USED_REF: str
    BAM_EDIT: str
    GENE_PHENO: str
    SIFT: str
    PolyPhen: str
    EXON: str
    INTRON: str
    DOMAINS: str
    miRNA: str
    HGVSc: str
    HGVSp: str
    HGVS_OFFSET: str
    AF: str
    AFR_AF: str
    AMR_AF: str
    EAS_AF: str
    EUR_AF: str
    SAS_AF: str
    AA_AF: str
    EA_AF: str
    gnomAD_AF: str
    gnomAD_AFR_AF: str
    gnomAD_AMR_AF: str
    gnomAD_ASJ_AF: str
    gnomAD_EAS_AF: str
    gnomAD_FIN_AF: str
    gnomAD_NFE_AF: str
    gnomAD_OTH_AF: str
    gnomAD_SAS_AF: str
    MAX_AF: str
    MAX_AF_POPS: str
    CLIN_SIG: str
    SOMATIC: str
    PHENO: str
    PUBMED: str
    MOTIF_NAME: str
    MOTIF_POS: str
    HIGH_INF_POS: str
    MOTIF_SCORE_CHANGE: str
    TRANSCRIPTION_FACTORS: str
